ttnn/cpp/ttnn/operations/moreh/moreh_adamw/device/moreh_adamw_device_operation.cpp

```diff
@@ -0,0 +1,149 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "moreh_adamw_device_operation.hpp"
+
+#include "tt_dnn/op_library/moreh_helper_functions.hpp"
+#include "ttnn/tensor/types.hpp"
+
+namespace ttnn::operations::moreh::moreh_adamw {
+
+MorehAdamWDeviceOperation::program_factory_t MorehAdamWDeviceOperation::select_program_factory(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    return MultiCore{};
+}
+
+void MorehAdamWDeviceOperation::validate_inputs(
+    const operation_attributes_t& attributes, const tensor_args_t& tensor_args) {
+    tt::operations::primary::check_tensor(tensor_args.param_in, "moreh_adamw", "param_in");
+    tt::operations::primary::check_tensor(tensor_args.grad, "moreh_adamw", "grad");
+    tt::operations::primary::check_tensor(tensor_args.exp_avg_in, "moreh_adamw", "exp_avg_in");
+    tt::operations::primary::check_tensor(tensor_args.exp_avg_sq_in, "moreh_adamw", "exp_avg_sq_in");
+
+    if (tensor_args.max_exp_avg_sq_in.has_value()) {
+        tt::operations::primary::check_tensor(
+            tensor_args.max_exp_avg_sq_in.value(), "moreh_adamw", "max_exp_avg_sq_in");
+    }
+
+    if (tensor_args.param_out.has_value()) {
+        tt::operations::primary::check_tensor(tensor_args.param_out.value(), "moreh_adamw", "param_out");
+    }
+    if (tensor_args.exp_avg_out.has_value()) {
+        tt::operations::primary::check_tensor(tensor_args.exp_avg_out.value(), "moreh_adamw", "exp_avg_out");
+    }
+    if (tensor_args.exp_avg_sq_out.has_value()) {
+        tt::operations::primary::check_tensor(tensor_args.exp_avg_sq_out.value(), "moreh_adamw", "exp_avg_sq_out");
+    }
+    if (tensor_args.max_exp_avg_sq_out.has_value()) {
+        tt::operations::primary::check_tensor(
+            tensor_args.max_exp_avg_sq_out.value(), "moreh_adamw", "max_exp_avg_sq_out");
+    }
+}
+
+void MorehAdamWDeviceOperation::validate_on_program_cache_miss(
+    const operation_attributes_t& attributes, const tensor_args_t& tensor_args) {
+    validate_inputs(attributes, tensor_args);
+}
+
+void MorehAdamWDeviceOperation::validate_on_program_cache_hit(
+    const operation_attributes_t& attributes, const tensor_args_t& tensor_args) {
+    validate_inputs(attributes, tensor_args);
+}
+
+MorehAdamWDeviceOperation::shape_return_value_t MorehAdamWDeviceOperation::compute_output_shapes(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    auto input_tensor_shape = tensor_args.param_in.get_shape();
+
+    return {
+        input_tensor_shape,
+        input_tensor_shape,
+        input_tensor_shape,
+        input_tensor_shape,
+    };
+}
+
+MorehAdamWDeviceOperation::tensor_return_value_t MorehAdamWDeviceOperation::create_output_tensors(
```

Comment:
`using tensor_return_value_t = std::vector<std::optional<Tensor>>;`

But this method never adds nullopt in. Please change return type or change the implementation of this method.

Reply:
This should not have a different behavior from
https://github.com/tenstorrent/tt-metal/pull/12258/files#diff-fb497ecad3d50fd345470a12c2d384db1fee628713102f610e1d054e0225e6dfR84

Reply:
I've updated this func to added nullopt if amsgrad is false.